osp/core/cuds.py

```diff
@@ -335,93 +567,896 @@ def update(self, *args: "Cuds") -> List["Cuds"]:
+                        values: Iterable[RDFCompatibleType]):
+        """Replace values assigned to a datatype property.
+
+        Args:
+            attribute: The ontology attribute to be used for assignments.
+            values: An iterable of Python types that are compatible either
+                with the OWL standard's data types for literals or compatible
+                with OSP-core as custom data types.
+
+        Raises:
+            TypeError: When Python objects with types incompatible with the
+                OWL standard or with OSP-core as custom datatypes are given.
+        """
+        # TODO: prevent the end result having more than one value than one
+        #  depending on ontology cardinality restrictions and/or functional
+        #  property criteria.
+        values = set(values)
+        for x in values:
+            if not isinstance(x, RDF_COMPATIBLE_TYPES):
+                logger.warning(f"Type '{type(x)}' of object {x} cannot "
+                               f"be set as attribute value, as it is "
+                               f"incompatible with the OWL standard")
+        if self.session:
+            self.session._notify_read(self)
+        self._graph.remove((self.iri, attribute.iri, None))
+        for value in values:
+            self._graph.add((self.iri, attribute.iri,
+                             Literal(attribute.convert_to_datatype(value),
+                                     datatype=attribute.datatype)))
+        if self.session:
+            self.session._notify_update(self)
+
+    def _attribute_value_generator(self,
+                                   attribute: OntologyAttribute,
+                                   _notify_read: bool = True) \
+            -> Iterator[RDFCompatibleType]:
+        """Returns a generator of values assigned to the specified attribute.
+
+        Args:
+            attribute: The ontology attribute query for values.
 
         Returns:
-            Iterator[Cuds]: The queried objects.
+            Generator that returns the attribute values.
         """
-        if return_rel:
-            collected_uids, mapping = self._get(*uids, rel=rel, oclass=oclass,
-                                                return_mapping=True)
+        # TODO (detach cuds from sessions): Workaround to keep the behavior:
+        #  removed CUDS do not have attributes. Think of a better way to
+        #  detach CUDS from sessions. `self._graph is not
+        #  self.session.graph` happens when `session._notify_read` is called
+        #  for this cuds, but this is hacky maybe not valid in general for
+        #  all sessions.
+        if self.session is None or\
+                self._graph is not self.session.graph:
+            raise AttributeError(f"The CUDS {self} does not belong to any "
+                                 f"session. None of its attributes are "
+                                 f"accessible.")
+
+        if _notify_read and self.session:
```

Comment:
It might be the case that other part of the code expects an `AttributeError` here when the session is none. I'll try the tests again without the check and see.

Reply:
I mean in line 961, the `and self.session`. If it was `False` (as in `None`), the code would have stopped in the previous check.